### DISCUSSION TITLE
compile abort win 32 and 64 bit versions with Ubuntu

### DISCUSSION THREAD

**Discussion by JohnnyR8** [diasurgical/DevilutionX#7445] [Q&A]
building Linux devX 1.6.0 works perfect, but when i try to compile the windows 32 and 64 bit versions with Ubuntu i get this error at 87 % and abort - any ideas?

cmake --build build -j $(getconf _NPROCESSORS_ONLN)
.
.
[ 87%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/trn.cpp.obj
[ 87%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/render/automap_render.cpp.obj
gmake[2]: *** [Source/CMakeFiles/libdevilutionx.dir/build.make:1307: Source/CMakeFiles/libdevilutionx.dir/dvlnet/abstract_net.cpp.obj] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:3063: Source/CMakeFiles/libdevilutionx.dir/all] Error 2
gmake: *** [Makefile:166: all] Error 2

**Comment by Trihedraf** [comment]
You need to include more of the log. That doesn't show anything other than there is an error and it is stopping the compile.

**Reply by JohnnyR8** [reply]
you mean more lines above the abort or is there a Terminal setting to show more?

**Reply by glebm** [reply]
More lines prior to the lines you posted

**Comment by JohnnyR8** [comment]
[ 69%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine.cpp.obj
[ 69%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/gmenu.cpp.obj
[ 69%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/help.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/hwcursor.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/init.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/interfac.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/inv.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/itemdat.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/items.cpp.obj
In file included from /home/ubuntu/devilutionx/Source/init.cpp:16:
/home/ubuntu/devilutionx/build/_deps/find_steam_game-src/find_steam_game.h:47: warning: "NOMINMAX" redefined
   47 | #define NOMINMAX
      | 
In file included from /usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/x86_64-w64-mingw32/bits/c++config.h:679,
                 from /usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/type_traits:38,
                 from /usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/optional:37,
                 from /home/ubuntu/devilutionx/Source/init.h:8,
                 from /home/ubuntu/devilutionx/Source/init.cpp:6:
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/x86_64-w64-mingw32/bits/os_defines.h:45: note: this is the location of the previous definition
   45 | #define NOMINMAX 1
      | 
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/lighting.cpp.obj
[ 71%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/loadsave.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/menu.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/minitext.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/misdat.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/missiles.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/monstdat.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/monster.cpp.obj
[ 72%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/movie.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/msg.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/multi.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/nthread.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/objdat.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/objects.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/options.cpp.obj
[ 73%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/pack.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/pfile.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/player.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/playerdat.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/plrmsg.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/portal.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/quests.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/restrict.cpp.obj
[ 74%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/spelldat.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/spells.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/stores.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/sync.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/textdat.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/tmsg.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/towners.cpp.obj
[ 75%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/track.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/axis_direction.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/controller.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/controller_buttons.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/controller_motion.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/devices/joystick.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/devices/kbcontroller.cpp.obj
[ 77%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/game_controls.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/menu_controls.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/modifier_hints.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/controls/plrctrls.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/data/file.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/data/parser.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/data/record_reader.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/button.cpp.obj
[ 78%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/credits.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/credits_lines.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/diabloui.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/dialogs.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/hero/selhero.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/mainmenu.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/multi/selconn.cpp.obj
[ 79%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/multi/selgame.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/progress.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/scrollbar.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/selok.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/selstart.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/selyesno.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/settingsmenu.cpp.obj
[ 80%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/support_lines.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/title.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/DiabloUI/text_input.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/abstract_net.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/base.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/cdwrap.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/frame_queue.cpp.obj
[ 81%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/loopback.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/dvlnet/packet.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/actor_position.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/animationinfo.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/assets.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/backbuffer_state.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/direction.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/dx.cpp.obj
[ 83%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/events.cpp.obj
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/load_cel.cpp.obj
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/load_cl2.cpp.obj
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/load_clx.cpp.obj
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/load_pcx.cpp.obj
In file included from /home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.condition_variable.h:52,
                 from /home/ubuntu/devilutionx/build/_deps/mingw-std-threads-build/cmake_stdheaders_generator/cmake_stdheaders_generator/condition_variable:10,
                 from /usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/mutex:48,
                 from /home/ubuntu/devilutionx/build/_deps/mingw-std-threads-build/cmake_stdheaders_generator/cmake_stdheaders_generator/mutex:9,
                 from /usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:40,
                 from /home/ubuntu/devilutionx/build/_deps/mingw-std-threads-build/cmake_stdheaders_generator/cmake_stdheaders_generator/future:9,
                 from /home/ubuntu/devilutionx/build/_deps/asio-src/asio/include/asio/detail/future.hpp:20,
                 from /home/ubuntu/devilutionx/build/_deps/asio-src/asio/include/asio/packaged_task.hpp:19,
                 from /home/ubuntu/devilutionx/build/_deps/asio-src/asio/include/asio/ts/executor.hpp:32,
                 from /home/ubuntu/devilutionx/build/_deps/asio-src/asio/include/asio/ts/net.hpp:19,
                 from /home/ubuntu/devilutionx/Source/dvlnet/tcp_client.h:18,
                 from /home/ubuntu/devilutionx/Source/dvlnet/abstract_net.cpp:15:
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.mutex.h:483:17: note: ‘#pragma message: This version of MinGW seems to include a win32 port of pthreads, and probably already has C++11 std threading classes implemented, based on pthreads. These classes, found in namespace std, are not overridden by the mingw-std-thread library. If you would still like to use this implementation (as it is more lightweight), use the classes provided in namespace mingw_stdthread.’
  483 | #pragma message "This version of MinGW seems to include a win32 port of\
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  484 |  pthreads, and probably already has C++11 std threading classes implemented,\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  485 |  based on pthreads. These classes, found in namespace std, are not overridden\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  486 |  by the mingw-std-thread library. If you would still like to use this\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  487 |  implementation (as it is more lightweight), use the classes provided in\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  488 |  namespace mingw_stdthread."
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/palette.cpp.obj
In file included from /home/ubuntu/devilutionx/build/_deps/mingw-std-threads-build/cmake_stdheaders_generator/cmake_stdheaders_generator/future:10:
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:355:17: error: ‘__async_result_of’ was not declared in this scope; did you mean ‘std::__async_result_of’?
  355 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                 ^~~~~~~~~~~~~~~~~
      |                 std::__async_result_of
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:197:11: note: ‘std::__async_result_of’ declared here
  197 |     using __async_result_of = typename __invoke_result<
      |           ^~~~~~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:355:48: error: template argument 1 is invalid
  355 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                                                ^~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:702:17: error: ‘__async_result_of’ was not declared in this scope; did you mean ‘std::__async_result_of’?
  702 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                 ^~~~~~~~~~~~~~~~~
      |                 std::__async_result_of
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:197:11: note: ‘std::__async_result_of’ declared here
  197 |     using __async_result_of = typename __invoke_result<
      |           ^~~~~~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:702:48: error: template argument 1 is invalid
  702 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                                                ^~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:805:17: error: ‘__async_result_of’ was not declared in this scope; did you mean ‘std::__async_result_of’?
  805 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                 ^~~~~~~~~~~~~~~~~
      |                 std::__async_result_of
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:197:11: note: ‘std::__async_result_of’ declared here
  197 |     using __async_result_of = typename __invoke_result<
      |           ^~~~~~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:805:48: error: template argument 1 is invalid
  805 |   friend future<__async_result_of<_Fn, _Args...>> async (std::launch, _Fn &&, _Args&&...);
      |                                                ^~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:848:18: error: ‘class mingw_stdthread::future<mingw_stdthread::detail::Empty> mingw_stdthread::future<mingw_stdthread::detail::Empty>::future’ is private within this context
  848 |   shared_future (future<void> && source) noexcept
      |                  ^~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:416:7: note: declared private here
  416 | class shared_future : future<T>
      |       ^~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:853:36: error: ‘class mingw_stdthread::future<mingw_stdthread::detail::Empty> mingw_stdthread::future<mingw_stdthread::detail::Empty>::future’ is private within this context
  853 |   shared_future<void> & operator= (future<void> && source) noexcept
      |                                    ^~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:416:7: note: declared private here
  416 | class shared_future : future<T>
      |       ^~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:924:8: error: ‘uses_allocator’ is not a class template
  924 | struct uses_allocator<promise<T>, Alloc> : std::true_type
      |        ^~~~~~~~~~~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:1025:3: error: redefinition of ‘template<class Function, class ... Args> std::future<typename std::__invoke_result<typename std::decay<_Tp>::type, typename std::decay<_Args>::type ...>::type> std::async(Function&&, Args&& ...)’
 1025 |   async(Function&& f, Args&&... args)
      |   ^~~~~
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:1828:5: note: ‘template<class _Fn, class ... _Args> std::future<typename std::__invoke_result<typename std::decay<_Tp>::type, typename std::decay<_Args>::type ...>::type> std::async(_Fn&&, _Args&& ...)’ previously declared here
 1828 |     async(_Fn&& __fn, _Args&&... __args)
      |     ^~~~~
/home/ubuntu/devilutionx/build/_deps/mingw-std-threads-src/mingw.future.h:1042:3: error: redefinition of ‘template<class Function, class ... Args> std::future<typename std::__invoke_result<typename std::decay<_Tp>::type, typename std::decay<_Args>::type ...>::type> std::async(launch, Function&&, Args&& ...)’
 1042 |   async(std::launch policy, Function&& f, Args&&... args)
      |   ^~~~~
/usr/lib/gcc/x86_64-w64-mingw32/13-win32/include/c++/future:1794:5: note: ‘template<class _Fn, class ... _Args> std::future<typename std::__invoke_result<typename std::decay<_Tp>::type, typename std::decay<_Args>::type ...>::type> std::async(launch, _Fn&&, _Args&& ...)’ previously declared here
 1794 |     async(launch __policy, _Fn&& __fn, _Args&&... __args)
      |     ^~~~~
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/path.cpp.obj
[ 84%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/random.cpp.obj
[ 85%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/sound_position.cpp.obj
[ 85%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/surface.cpp.obj
[ 85%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/trn.cpp.obj
[ 85%] Building CXX object Source/CMakeFiles/libdevilutionx.dir/engine/render/automap_render.cpp.obj
gmake[2]: *** [Source/CMakeFiles/libdevilutionx.dir/build.make:1307: Source/CMakeFiles/libdevilutionx.dir/dvlnet/abstract_net.cpp.obj] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [CMakeFiles/Makefile2:2476: Source/CMakeFiles/libdevilutionx.dir/all] Error 2
gmake: *** [Makefile:146: all] Error 2

**Reply by AJenbo** [reply]
The version of gcc you are using has issues with threading on windows, use an older one and it should work

**Reply by StephenCWills** [reply]
I got around this the other day by adding `-DDISABLE_ZERO_TIER` to the CMake configuration. If you don't need Zero Tier, this could also work for you.

**Reply by JohnnyR8** [reply]
thank you for your replies :-)

i tried it with gcc 10,11,12 and 13 but had no luck.

where exactly is CMake configuration located to add  -DDISABLE_ZERO_TIER?

**Reply by AJenbo** [reply]
Can you give the output of the cmake config command so we can check that you are acutally using GCC 10? Installling it is not enough if it's still configured to use 13.

**Reply by JohnnyR8** [reply]
ubuntu@ubuntu:~/devilutionx$ gcc --version
gcc (Ubuntu 10.5.0-4ubuntu2) 10.5.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

**Reply by AJenbo** [reply]
That confirms that you have version 10 installed and set as the default, but it's not the cmake config output.

**Reply by StephenCWills** [reply]
> where exactly is CMake configuration located to add -DDISABLE_ZERO_TIER?

Assuming you're following our build instructions, add it to the end of this command...

https://github.com/diasurgical/devilutionX/blob/1078598226b0b154415ffca51ab29b619cbd04e6/docs/building.md?plain=1#L204

Like so...

```
cmake -S. -Bbuild -DCMAKE_TOOLCHAIN_FILE=../CMake/platforms/mingwcc.toolchain.cmake -DCMAKE_BUILD_TYPE=Release -DDEVILUTIONX_SYSTEM_BZIP2=OFF -DDISABLE_ZERO_TIER=ON
```

---

If you're referring to the CI scripts, add it to the end of this command...

https://github.com/diasurgical/devilutionX/blob/1078598226b0b154415ffca51ab29b619cbd04e6/.github/workflows/Windows_MinGW_x64.yml#L39

Like so...

```
cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_TESTING=OFF -DCPACK=ON -DCMAKE_TOOLCHAIN_FILE=../CMake/platforms/mingwcc64.toolchain.cmake -DDEVILUTIONX_SYSTEM_BZIP2=OFF -DDEVILUTIONX_STATIC_LIBSODIUM=ON -DDISCORD_INTEGRATION=ON -DSCREEN_READER_INTEGRATION=ON -DDISABLE_ZERO_TIER=ON
```

**Reply by StephenCWills** [reply]
> That confirms that you have version 10 installed and set as the default, but it's not the cmake config output.

In his case, CMake should be using MinGW, so he would have to install an older version of `x86_64-w64-mingw32-gcc`, not `gcc`.

**Reply by JohnnyR8** [reply]
thank you Stephen - it worked!

but i am always interested in learning so Ajenbo, please let me know the cmake config command to use GCC 10

**Reply by AJenbo** [reply]
Like Stephen said you need the mingw-gcc, not system (linux) one.

**Comment by JohnnyR8** [comment]
- i tried no to run devolutionx.exe, 32 and 64 bit versions under Windows but it asks for: SDL2.dll, libsodium-26.dll and zlib1.dll.
  how do i get them?


- also tried to install an older version of mingw-gcc like described under this: [https://askubuntu.com/questions/1353046/cant-install-specific-version-of-gcc-mingw](https://askubuntu.com/questions/1353046/cant-install-specific-version-of-gcc-mingw)
but when i then compile devX 1.6 (cmake --build build -j $(getconf _NPROCESSORS_ONLN)) it still tooks the mingw gcc 13 version and crashes... how can i force it to use my new installed mingw gcc?

**Reply by qndel** [reply]
you get dlls from google or from devilutionx release package

**Reply by StephenCWills** [reply]
> i tried no to run devolutionx.exe, 32 and 64 bit versions under Windows but it asks for: SDL2.dll, libsodium-26.dll and zlib1.dll.
>   how do i get them?

The dlls don't get copied to the build folder alongside the exe. `CPACK` has to get them from another location to pull them together into a release package.

https://github.com/diasurgical/devilutionX/blob/1078598226b0b154415ffca51ab29b619cbd04e6/CMakeLists.txt#L531-L533

If you configured CMake with `-DCPACK=ON`, then you should be able to build the `package` target to have `CPACK` create a zip file for you that includes all the dlls.

https://github.com/diasurgical/devilutionX/blob/1078598226b0b154415ffca51ab29b619cbd04e6/.github/workflows/Windows_MinGW_x64.yml#L44

---

> how can i force it to use my new installed mingw gcc?

These lines of code in the toolchain file determine which compilers CMake will use.

https://github.com/diasurgical/devilutionX/blob/1078598226b0b154415ffca51ab29b619cbd04e6/CMake/platforms/mingwcc64.toolchain.cmake#L10-L12

**Reply by AJenbo** [reply]
Alternatively you can simply uninstall the mingw gcc that doesn't work

**Reply by AJenbo** [reply]
You probably need the extra dll files because you are now also building using Visual Studio, please let us know what you are doing so we have a fair chance of answering your questions and and don't have to combine details from multiple posts to figure out what is going on.

**Reply by JohnnyR8** [reply]
yes i also build it with Visual Studio under windows, but in this discussion i stay by ubuntu (building for windows) :-)

**Reply by AJenbo** [reply]
Was the one that you said needed dll's build on Ubuntu?

**Reply by StephenCWills** [reply]
@AJenbo The build folder looks something like this when it's done.

![image](https://github.com/user-attachments/assets/6c3fc172-6eff-4d1e-a990-254918968a94)

If I try to run it, I get this.

![image](https://github.com/user-attachments/assets/1b32f14b-ca85-49de-82da-f5d787f242f3)

`CPACK` pulls the dlls out of the aforementioned `SDL2_WIN32_DLLS_DIR` folder. The folder it preps for the release looks more like this.

![image](https://github.com/user-attachments/assets/b2e563d8-4345-4605-a488-c86e92fc3fae)

Though, keep in mind that this is from my messy build system. Normally, you shouldn't have two versions of libsodium, for instance.

**Reply by AJenbo** [reply]
Or ttf or freetype 🙂

I didn't expect zlib1.dll which is what confuses me

**Reply by JohnnyR8** [reply]
> Was the one that you said needed dll's build on Ubuntu?

yes

**Reply by AJenbo** [reply]
Oh, well in that case on solution could be to download a test build or release and take the DLLs from there, hop that helps :)

**Reply by JohnnyR8** [reply]
> Alternatively you can simply uninstall the mingw gcc that doesn't work

i tried uninstalling with: sudo apt-get remove gcc-mingw-w64

but dont think this syntax for uninstalling is right... do you know it?

also is there something similar to manage multiple mingw gcc as for gcc ("update-alternatives tool") ?
https://phoenixnap.com/kb/install-gcc-ubuntu


installing another with: sudo apt install gcc-mingw-w64=9.3.0-7ubuntu1+22~exp1ubuntu4 -s -t focal 
is correct?
